crates/walrus-service/src/node/storage/blob_info.rs

```diff
@@ -149,6 +161,120 @@ impl BlobInfoTable {
                 [(object_id, per_object_operation.to_bytes())],
             )?;
         }
+
+        batch.insert_batch(&latest_handled_event_index, [(&(), event_index)])?;
+        batch.write()
+    }
+
+    /// Updates the blob info for a blob based on the [`BlobEvent`] when the node is in recovery
+    /// with incomplete history.
+    ///
+    /// Only updates the info if the provided `event_index` hasn't been processed yet.
+    #[tracing::instrument(skip(self))]
+    pub fn update_blob_info_during_recovery_with_incomplete_history(
+        &self,
+        event_index: u64,
+        event: &BlobEvent,
+        first_epoch: Epoch,
+        epoch_at_start: Epoch,
+    ) -> Result<(), TypedStoreError> {
+        tracing::debug!("updating blob info during recovery with incomplete history");
+        let extension_event = match event {
+            BlobEvent::Registered(BlobRegistered { end_epoch, .. })
+            | BlobEvent::Certified(BlobCertified { end_epoch, .. })
+            | BlobEvent::Deleted(BlobDeleted { end_epoch, .. })
+                if end_epoch <= &epoch_at_start =>
+            {
+                tracing::debug!(
+                    "skip updating blob info for event with end epoch before epoch at start"
+                );
+                return Ok(());
+            }
+            BlobEvent::Registered(_)
+            | BlobEvent::Certified(BlobCertified {
+                is_extension: false,
+                ..
+            })
+            | BlobEvent::Deleted(_)
+            | BlobEvent::InvalidBlobID(_)
+            | BlobEvent::DenyListBlobDeleted(_) => {
+                tracing::debug!("performing standard blob-info update for event");
+                return self.update_blob_info(event_index, event);
+            }
+            BlobEvent::Certified(event) => {
+                // Extensions need special handling.
+                event.clone()
+            }
```

Comment:
Suggestion (in a follow up PR): I think it worth adding an explicit test the case where certified event with earlier end_epoch is in an expired event blob, but extension event with later end_epoch is not. I'm not sure if the simtest, where extension is issued immediately after certification will exercise this code path.

Reply:
Created a follow-up issue and added a comment.